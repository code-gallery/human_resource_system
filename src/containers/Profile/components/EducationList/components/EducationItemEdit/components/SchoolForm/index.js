export { default, updateObjectInArray } from './SchoolForm'
