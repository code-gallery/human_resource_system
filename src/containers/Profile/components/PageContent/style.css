.PageContent .profile-experience-block .Avatar {
  float: left;
}

@media (max-width: 768px) {
  .PageContent .profile-experience-wrapper {
    padding: 10px;
  }

  .PageContent .profile-experience-block .Avatar {
    padding-bottom: 38px;
  }

  .PageContent .profile-experience-block .Avatar img {
    width: 104px;
    height: 104px;
    margin-right: 10px;
  }

  .PageContent .visible-xs .profile-status-ico {
    position: absolute;
    top: 114px;
    width: 30px;
    height: 30px;
    background-size: 22px 22px;
    left: 44px;
  }
}
