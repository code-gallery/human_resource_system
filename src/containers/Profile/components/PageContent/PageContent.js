import React, { Component } from 'react'
import PropTypes from 'prop-types'
import _keys from 'lodash/keys'
import DragSortableList from 'react-drag-sortable'
import { DEFAULT_ORDER } from 'containers/constants'
import EducationList from '../EducationList'
import WorkExperience from '../WorkExperience'
import AwardList from '../AwardList'
import RightBlock from '../OverviewBlock'
import EditProfile from '../OverviewBlock/components/EditProfile'
import './style.css'

function defaultOrder(pageOrder) {
  if (Object.keys(pageOrder).length < 9) {
    return DEFAULT_ORDER
  } else {
    return pageOrder
  }
}

class PageContent extends Component {
  constructor(props) {
    super(props)
    const pageOrder = props.pageOrder ? props.pageOrder : {}
    this.state = {
      reorder: false,
      order: defaultOrder(pageOrder)
    }
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.pageOrder !== nextProps.pageOrder &&
      _keys(nextProps.pageOrder).length !== 0) {
      this.setState({
        order: nextProps.pageOrder
      })
    }
    if (this.props.editMode !== nextProps.editMode) {
      this.setState({
        reorder: false
      })
    }
  }

  toggleReorder = () => {
    this.setState({
      reorder: !this.state.reorder
    })
  }

  onSort = (sortedList) => {
    const _obj = {}
    sortedList.map((item) => {
      _obj[item.classes[0]] = item.rank
      return true
    })
    this.props.onChangeProfileField('profile_order', _obj)
    return true
  }

  renderComponents() {
    const { reorder, order } = this.state
    const {
      editMode, educations, jobs,gaps, reference,
      deleteWork, deleteEducation, deleteAward
    } = this.props
    const { degrees, awards, industries, highSchoolDegrees, gapsReasons } = reference
    const {
      user,
      allAwards: {
        award,
        cpd,
        language,
        certificate,
        achievement,
        project,
        skill
      }
    } = this.props
    const elements = {
      education: (<EducationList
        education={educations}
        editMode={editMode}
        awards={awards}
        highSchoolDegrees={highSchoolDegrees}
        qualifications={degrees}
        reorder={reorder}
        toggleReorder={this.toggleReorder}
        deleteEducation={deleteEducation}
        saveEntity={this.props.saveEntity}
        filterEntries={this.props.filterEntries}
        user={user}
      />),
      jobs: (<WorkExperience
        jobs={jobs}
        editMode={editMode}
        industries={industries}
        gapsReasons={gapsReasons}
        reorder={reorder}
        deleteWork={deleteWork}
        toggleReorder={this.toggleReorder}
        saveEntity={this.props.saveEntity}
        filterEntries={this.props.filterEntries}
        user={user}
      />),
      award: (<AwardList
        awardType="award"
        awardTitle="Awards"
        awardList={award}
        editMode={editMode}
        reorder={reorder}
        toggleReorder={this.toggleReorder}
        deleteAward={deleteAward}
        saveAward={this.props.saveAward}
        filterEntries={this.props.filterEntries}
        user={user}
      />),
      cpd: (<AwardList
        awardType="cpd"
        awardTitle="Professional Development (CPD)"
        awardList={cpd}
        editMode={editMode}
        reorder={reorder}
        toggleReorder={this.toggleReorder}
        deleteAward={deleteAward}
        saveAward={this.props.saveAward}
        filterEntries={this.props.filterEntries}
        user={user}
      />),
      certificate: (<AwardList
        awardType="certificate"
        awardTitle="Certificates"
        awardList={certificate}
        editMode={editMode}
        reorder={reorder}
        toggleReorder={this.toggleReorder}
        deleteAward={deleteAward}
        saveAward={this.props.saveAward}
        filterEntries={this.props.filterEntries}
        user={user}
      />),
      achievement: (<AwardList
        awardType="achievement"
        awardTitle="Achievements"
        awardList={achievement}
        editMode={editMode}
        reorder={reorder}
        toggleReorder={this.toggleReorder}
        deleteAward={deleteAward}
        saveAward={this.props.saveAward}
        filterEntries={this.props.filterEntries}
        user={user}
      />),
      project: (<AwardList
        awardType="project"
        awardTitle="Projects"
        awardList={project}
        editMode={editMode}
        reorder={reorder}
        toggleReorder={this.toggleReorder}
        deleteAward={deleteAward}
        saveAward={this.props.saveAward}
        filterEntries={this.props.filterEntries}
        user={user}
      />),
      skill: (<AwardList
        awardType="skill"
        awardTitle="Skills"
        awardList={skill}
        editMode={editMode}
        reorder={reorder}
        toggleReorder={this.toggleReorder}
        deleteAward={deleteAward}
        saveAward={this.props.saveAward}
        filterEntries={this.props.filterEntries}
        user={user}
      />),
      language: (<AwardList
        awardType="language"
        awardTitle="Languages"
        awardList={language}
        editMode={editMode}
        reorder={reorder}
        toggleReorder={this.toggleReorder}
        deleteAward={deleteAward}
        saveAward={this.props.saveAward}
        filterEntries={this.props.filterEntries}
        user={user}
      />)
    }
    const _order = Object.keys(order).sort(function (a, b) {
      return order[a] - order[b]
    })
    const list = _order.map((key) => {
      return { content: (elements[key]), classes: [ key ] }
    })

    return (
      <DragSortableList
        items={list}
        moveTransitionDuration={1}
        dropBackTransitionDuration={0.3}
        onSort={this.onSort}
        type="vertical"
        disableDrag={!(reorder && editMode)}
      />
    )
  }

  render() {
    const { onChangeProfileField, allAwards, jobs,  educations, editMode, user, activities } = this.props
    const invalidFields = []
    if (user.first_name === '') {
      invalidFields.push('first_name')
    }
    if (user.last_name === '') {
      invalidFields.push('last_name')
    }

    return (
      <div className="container PageContent">
        <div className="row">
          {editMode &&
            <div className="col-xs-12 visible-xs visible-sm">
              <EditProfile
                onChangeProfileField={onChangeProfileField}
                user={user}
                invalidFields={invalidFields}
              />
            </div>
          }
          <div className="col-xs-12 col-md-8">
            {this.renderComponents()}
          </div>
          <RightBlock
            allAwards={allAwards}
            jobs={jobs}
            educations={educations}
            editMode={editMode}
            user={user}
            onChangeProfileField={onChangeProfileField}
            activities={activities}
          />
        </div>
      </div>
    )
  }
}

PageContent.propTypes = {
  pageOrder: PropTypes.object,
  user: PropTypes.object,
  editMode: PropTypes.bool,
  allAwards: PropTypes.object,
  educations: PropTypes.array,
  activities: PropTypes.array,
  jobs: PropTypes.array,
  reference: PropTypes.shape({
    awards: PropTypes.array,
    degrees: PropTypes.array,
    industries: PropTypes.array,
    gapsReasons: PropTypes.array
  }),
  onChangeProfileField: PropTypes.func.isRequired,
  deleteAward: PropTypes.func.isRequired,
  deleteWork: PropTypes.func.isRequired,
  deleteEducation: PropTypes.func.isRequired,
  saveAward: PropTypes.func.isRequired,
  saveEntity: PropTypes.func.isRequired,
  filterEntries: PropTypes.func.isRequired
}

export default PageContent
