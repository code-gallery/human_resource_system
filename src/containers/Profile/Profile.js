import React from 'react'
import PropTypes from 'prop-types'
import Notifications, { notify } from 'react-notify-toast'
import Layout from 'containers/Layout'
import { NOTIFICATION_TIMEOUT } from 'containers/constants'
import ContentCard from 'components/ContentCard'
import LoadingIndicator from 'components/LoadingIndicator'
import { HeroBlock, ProfileInfo } from 'components/Profile'
import AppBanner from 'components/Abstract/AppBanner'
import PageContent from './components/PageContent'

class Profile extends AppBanner {
  constructor(props) {
    super(props)
    this.checkBiometricStatus = false
    this.saving = false
    this.state = {
      user: props.user
    }
  }

  componentWillMount() {
    const { uid } = this.props.match.params || ''
    if (uid) {
      // fetch profile data of another user
      this.checkBiometricStatus = parseInt(uid, 10) > 0
      this.props.fetchOtherUserProfile(uid)
    } else {
      // fetch your own profile
      this.props.fetchActivities()
      this.props.fetchUserProfile()
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.user !== this.props.user) {
      this.setState({
        user: nextProps.user
      })
      this.user = nextProps.user
    }

    if (this.saving) {
      if (this.props.saveErrorMsg !== nextProps.saveErrorMsg && nextProps.saveErrorMsg !== null) {
        // server error while saving
        this.saving = false
        this.notifyFn(nextProps.saveErrorMsg, 'error')
      } else if (this.props.pending && !nextProps.pending) {
        this.saving = false
        // saved successfully
        this.props.setEditMode(false)
        this.notifyFn('Profile was updated  successfully', 'success')
      }
    }
  }

  notifyFn(message, type) {
    // this is a wrapper function to make the unit test happy
    notify.show(message, type, NOTIFICATION_TIMEOUT)
  }

  saveChanges = () => {
    this.saving = true
    this.props.fetchUserProfile()
    const { user } = this.state
    if (user.first_name !== '' && user.last_name !== '') {
      this.props.updateUserProfile(user)
    } else {
      this.notifyFn('"First Name" and "Last Name" are mandatory', 'error')
    }
  }

  undoChanges = () => {
    this.setState({
      user: this.user
    })
  }

  deleteAward = (id, type) => {
    this.props.deleteAward({ id, type },
      () => notify.show(`Award "${type}" deleted successfully`, 'success', NOTIFICATION_TIMEOUT),
      () => notify.show(`Award "${type}" was not deleted. Please try again`, 'error', NOTIFICATION_TIMEOUT)
    )
  }

  deleteWork = (id) => {
    this.props.deleteJob({ id },
      () => notify.show('Work deleted successfully', 'success', NOTIFICATION_TIMEOUT),
      () => notify.show('Failed to delete. Please try again', 'error', NOTIFICATION_TIMEOUT)
    )
  }

  deleteEducation = (id) => {
    this.props.deleteEducation({ id },
      () => notify.show('Education deleted successfully', 'success', NOTIFICATION_TIMEOUT),
      () => notify.show('Failed to delete. Please try again', 'error', NOTIFICATION_TIMEOUT)
    )
  }

  genericError() {
    notify.show('Failed to save. Please try again', 'error', NOTIFICATION_TIMEOUT)
  }

  saveAward = (data, verify) => {
    if (data.id) {
      if (verify === 'verified') {
        this.props.editAward(data,
          () => {
            notify.show('Award was edited successfully.', 'success', NOTIFICATION_TIMEOUT)
          },
          this.genericError
        )
      } else {
        this.props.editAward(data,
          (award_data) => {
            if (verify) {
              this.props.verifyAward({ id: award_data.id, type: award_data.type })
            }
            notify.show('Award was edited successfully.', 'success', NOTIFICATION_TIMEOUT)
          },
          this.genericError
        )
      }
    } else {
      this.props.saveAward(data,
        (award_data) => {
          if (verify) {
            this.props.verifyAward({ id: award_data.id, type: award_data.type })
          }
          notify.show('Award was created successfully', 'success', NOTIFICATION_TIMEOUT)
        },
        this.genericError
      )
    }
  }

  saveEntity = (data, verify) => {
    this.props.fetchUserProfile()
    if (data.id) {
      if (verify === 'verified') {
        this.props.editEntity(data,
          () => {
            notify.show('Entity was edited successfully', 'success', NOTIFICATION_TIMEOUT)
          },
          this.genericError
        )
      } else {
        this.props.editEntity(data,
          (entity_data) => {
            if (verify) {
              this.props.verifyEntity({ id: entity_data.id, type: data.entity_type })
            }
            notify.show('Entity was edited successfully', 'success', NOTIFICATION_TIMEOUT)
          },
          this.genericError
        )
      }
    } else {
      this.props.saveEntity(data,
        (entity_data) => {
          if (verify) {
            this.props.verifyEntity({ id: entity_data.id, type: data.entity_type })
          }
          notify.show('New entity was created successfully', 'success', NOTIFICATION_TIMEOUT)
        },
        this.genericError
      )
    }
  }

  onImagePick = (evt, field) => {
    const image = evt.target.files[0]
    const reader = new FileReader()
    reader.readAsDataURL(image)
    reader.onload = (data) => {
      this.setState({
        user: {
          ...this.state.user,
          [field]: data.target.result
        }
      })
    }
  }

  updateProfileImage = (profile_image) => {
    this.setState({
      user: {
        ...this.state.user,
        profile_image
      }
    })
  }

  onChangeProfileField = (field, value) => {
    if (field === 'location') {
      const { formatted_address } = value
      const addressComponents = formatted_address.split(',')
      const town = addressComponents[0]
      const country = addressComponents[addressComponents.length - 1]
      this.setState({
        user: {
          ...this.state.user,
          town,
          country
        }
      })
      return
    }
    this.setState({
      user: {
        ...this.state.user,
        [field]: value
      }
    })
  }

  filterEntries = (item) => {
    const { uid } = this.props.match.params || ''
    if (uid) {
      if (item.verified_status === 'declined') {
        item.verified_status = 'not_verified'
      }
    }
    return true
  }

  setDisplayContact = () => {
    if (window.innerWidth > 986) {
      return true
    } else {
      return !this.props.isMobileNavActive
    }
  }

  renderProfileNotFound() {
    return (
      <Layout>
        <ContentCard title="Profile not found" className="ContentCard-marginTop">
          <p>Unfortunately, we cannot find the profile you are looking for.</p>
        </ContentCard>
      </Layout>
    )
  }

  renderUnverifiedProfile() {
    return (
      <Layout>
        <ContentCard title="Unverified profile" className="ContentCard-marginTop">
          <p>Unfortunately, you cannot view this profile until it has been biometrically identified.</p>
        </ContentCard>
      </Layout>
    )
  }

  render() {
    const {
      pending,
      allAwards,
      jobs,
      educations,
      reference,
      editMode,
      setEditMode,
      activities,
      userProfileError
    } = this.props
    const { user } = this.state
    const { background_image, biometrics_status } = user
    const showAppBanner = user.first_name && !user.mobile_address

    if (pending) {
      return (
        <Layout>
          <Notifications />
          <LoadingIndicator size="80" />
        </Layout>
      )
    }

    if (this.checkBiometricStatus && biometrics_status !== 'complete') {
      // if url is /profile/:number and biometrics_status is not complete
      return this.renderUnverifiedProfile()
    }

    if (userProfileError !== '') {
      return this.renderProfileNotFound()
    }

    return (
      <Layout
        saveChanges={this.saveChanges}
        undoChanges={this.undoChanges}
      >
        <Notifications />
        { showAppBanner && super.render() }
        <HeroBlock
          background_image={background_image}
          onImagePick={this.onImagePick}
        />

        <ProfileInfo
          data={user}
          isProfile={true}
          editMode={editMode}
          onImagePick={this.onImagePick}
          updateProfileImage={this.updateProfileImage}
          allAwards={allAwards}
          educations={educations}
          jobs={jobs}
          displayContact={this.setDisplayContact()}
          setEditMode={setEditMode}
        />
        <PageContent
          user={user}
          allAwards={allAwards}
          educations={educations}
          jobs={jobs}
          reference={reference}
          editMode={editMode}
          pageOrder={user.profile_order}
          onChangeProfileField={this.onChangeProfileField}
          saveAward={this.saveAward}
          deleteAward={this.deleteAward}
          deleteWork={this.deleteWork}
          deleteEducation={this.deleteEducation}
          saveEntity={this.saveEntity}
          filterEntries={this.filterEntries}
          activities={activities}
        />
      </Layout>
    )
  }
}

Profile.defaultProps = {
  editMode: false
}

Profile.propTypes = {
  editMode: PropTypes.bool.isRequired,
  fetchUserProfile: PropTypes.func.isRequired,
  fetchOtherUserProfile: PropTypes.func.isRequired,
  fetchActivities: PropTypes.func.isRequired,
  updateUserProfile: PropTypes.func.isRequired,
  deleteAward: PropTypes.func.isRequired,
  deleteJob: PropTypes.func.isRequired,
  deleteEducation: PropTypes.func.isRequired,
  saveAward: PropTypes.func.isRequired,
  saveEntity: PropTypes.func.isRequired,
  editEntity: PropTypes.func.isRequired,
  editAward: PropTypes.func.isRequired,
  verifyAward: PropTypes.func.isRequired,
  verifyEntity: PropTypes.func.isRequired,
  match: PropTypes.shape({
    params: PropTypes.object.isRequired
  }),
  user: PropTypes.shape({
    background_image: PropTypes.string,
    biometrics_status: PropTypes.string,
    profile_order: PropTypes.object
  }),
  userProfileError: PropTypes.string.isRequired,
  allAwards: PropTypes.object,
  jobs: PropTypes.array,
  educations: PropTypes.array,
  pending: PropTypes.bool.isRequired,
  saveErrorMsg: PropTypes.oneOfType([
    PropTypes.object, PropTypes.string
  ]),
  reference: PropTypes.shape({
    awards: PropTypes.array,
    degrees: PropTypes.array,
    industries: PropTypes.array
  }),
  activities: PropTypes.array,
  setEditMode: PropTypes.func.isRequired,
  isMobileNavActive: PropTypes.bool
}

export default Profile
