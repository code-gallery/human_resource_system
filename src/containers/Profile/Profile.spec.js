import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'
import { BrowserRouter as Router } from 'react-router-dom'
import config from 'store'
import LoadingIndicator from 'components/LoadingIndicator'
import { HeroBlock, ProfileInfo } from 'components/Profile'
import PageContent from './components/PageContent'
import Profile from './Profile'

describe('containers/Profile', () => {
  const commonProps = {
    fetchOtherUserProfile: jest.fn(),
    fetchActivities: jest.fn(),
    fetchUserProfile: jest.fn(),
    deleteAward: jest.fn(),
    deleteJob: jest.fn(),
    deleteEducation: jest.fn(),
    updateUserProfile: jest.fn(),
    saveAward: jest.fn(),
    saveEntity: jest.fn(),
    editEntity: jest.fn(),
    editAward: jest.fn(),
    verifyAward: jest.fn(),
    verifyEntity: jest.fn(),
    setEditMode: jest.fn(),
    saveErrorMsg: null,
    editMode: false,
    pending: true,
    isMobileNavActive: false,
    userProfileError: ''
  }

  describe('when it is the current user profile', () => {
    let wrapper
    let props
    beforeEach(() => {
      props = {
        ...commonProps,
        match: {
          params: {}
        },
        user: {
          first_name: 'Jane',
          last_name: 'Doe'
        }
      }

      wrapper = shallow(
        <Profile {...props} />
      )
      jest.spyOn(wrapper.instance(), 'notifyFn').mockImplementation(function() {})
    })

    it('props.fetchUserProfile is called on "componentWillMount"', () => {
      expect(props.fetchUserProfile).toHaveBeenCalled()
    })

    describe('when data is loading', () => {
      it('renders <LoadingIndicator />', () => {
        expect(wrapper.find(LoadingIndicator).length).toEqual(1)
      })
    })

    describe('componentWillReceiveProps', () => {
      beforeEach(() => {
        props.setEditMode.mockReset()
      })

      describe('saving successfully', () => {
        it('setEditMode is called correctly', () => {
          expect(props.setEditMode).not.toHaveBeenCalled()
          wrapper.instance().saving = true
          wrapper.setProps({ pending: true })
          wrapper.setProps({ pending: false })
          expect(props.setEditMode).toHaveBeenCalledWith(false)
        })

        it('shows success notification', () => {
          wrapper.instance().saving = true
          wrapper.setProps({ pending: true })
          wrapper.setProps({ pending: false })
          expect(wrapper.instance().notifyFn).toHaveBeenCalledWith(expect.any(String), 'success')
        })
      })

      describe('server side error while saving', () => {
        beforeEach(() => {
          wrapper.instance().saving = true
          wrapper.setProps({ saveErrorMsg: 'error message' })
        })

        it('setEditMode is not called', () => {
          expect(props.setEditMode).not.toHaveBeenCalled()
        })

        it('shows error notification', () => {
          expect(wrapper.instance().notifyFn).toHaveBeenCalledWith(expect.any(String), 'error')
        })
      })

      describe('when user changes', () => {
        it('updates state correctly', () => {
          wrapper.setProps({
            user: {
              first_name: 'Jane Marguerite',
              last_name: 'Doe'
            }
          })
          expect(wrapper.state('user').first_name).toEqual('Jane Marguerite')
          expect(wrapper.instance().user.first_name).toEqual('Jane Marguerite')
        })
      })
    })

    describe('when data has loaded', () => {
      beforeEach(() => {
        wrapper.setProps({
          pending: false
        })
      })

      it('renders no <LoadingIndicator />', () => {
        expect(wrapper.find(LoadingIndicator).length).toEqual(0)
      })

      it('renders 1 <HeroBlock />', () => {
        expect(wrapper.find(HeroBlock).length).toEqual(1)
      })

      it('renders 1 <ProfileInfo />', () => {
        expect(wrapper.find(ProfileInfo).length).toEqual(1)
      })

      it('renders 1 <PageContent />', () => {
        expect(wrapper.find(PageContent).length).toEqual(1)
      })
    })

    describe('deleteAward', () => {
      it('calls props.deleteAward correctly', () => {
        wrapper.instance().deleteAward(126, 'language')
        expect(props.deleteAward).toHaveBeenCalledWith(
          { id: 126, type: 'language' },
          expect.any(Function),
          expect.any(Function)
        )
      })
    })

    describe('deleteWork', () => {
      it('calls props.deleteJob correctly', () => {
        wrapper.instance().deleteWork(126)
        expect(props.deleteJob).toHaveBeenCalledWith(
          { id: 126 },
          expect.any(Function),
          expect.any(Function)
        )
      })
    })

    describe('deleteEducation', () => {
      it('calls props.deleteEducation correctly', () => {
        wrapper.instance().deleteEducation(126)
        expect(props.deleteEducation).toHaveBeenCalledWith(
          { id: 126 },
          expect.any(Function),
          expect.any(Function)
        )
      })
    })

    describe('saveAward', () => {
      describe('when new (no id)', () => {
        it('calls props.saveAward correctly', () => {
          wrapper.instance().saveAward('data', false)
          expect(props.saveAward).toHaveBeenCalledWith(
            'data',
            expect.any(Function),
            expect.any(Function)
          )
        })
      })

      describe('when edit (id ok)', () => {
        it('calls props.editAward correctly', () => {
          const data = { id: 1, hello: 'world' }
          wrapper.instance().saveAward(data, false)
          expect(props.editAward).toHaveBeenCalledWith(
            data,
            expect.any(Function),
            expect.any(Function)
          )
        })
      })
    })

    describe('saveEntity', () => {
      describe('when new (no id)', () => {
        it('calls props.saveEntity correctly', () => {
          wrapper.instance().saveEntity('data', false)
          expect(props.saveEntity).toHaveBeenCalledWith(
            'data',
            expect.any(Function),
            expect.any(Function)
          )
        })
      })

      describe('when edit (id ok)', () => {
        it('calls props.editEntity correctly', () => {
          const data = { id: 1, hello: 'world' }
          wrapper.instance().saveEntity(data, false)
          expect(props.editEntity).toHaveBeenCalledWith(
            data,
            expect.any(Function),
            expect.any(Function)
          )
        })
      })
    })

    describe('updateProfileImage', () => {
      it('sets state correctly', () => {
        const expected = '/path/profile/image'
        wrapper.instance().updateProfileImage(expected)
        expect(wrapper.state().user).toEqual({
          ...props.user,
          profile_image: expected
        })
      })
    })

    describe('saveChanges', () => {
      describe('when first_name and last_name are valid', () => {
        it('calls props.updateUserProfile correctly', () => {
          wrapper.instance().saveChanges()
          expect(props.updateUserProfile).toHaveBeenCalledWith(props.user)
        })
      })

      describe('when first_name and last_name are not valid', () => {
        it('shows error notification', () => {
          wrapper.setState({
            user: {
              first_name: '',
              last_name: 'Doe'
            }
          })
          wrapper.instance().saveChanges()
          expect(wrapper.instance().notifyFn).toHaveBeenCalledWith(expect.any(String), 'error')
        })
      })
    })

    describe('undoChanges', () => {
      it('updates state correctly', () => {
        const user = {
          first_name: 'John',
          last_name: 'Smith',
          hello: 'world'
        }
        wrapper.instance().user = user
        wrapper.instance().undoChanges()
        expect(wrapper.state('user')).toEqual(user)
      })
    })

    describe('filterEntries', () => {
      describe('when uid is undefined', () => {
        it('returns correct value (1)', () => {
          const item = { id: 1, verified_status: 'declined' }
          wrapper.instance().filterEntries(item)
          expect(item.verified_status).toEqual('declined')
        })

        it('returns correct value (2)', () => {
          const item = { id: 1, verified_status: 'verified' }
          wrapper.instance().filterEntries(item)
          expect(item.verified_status).toEqual('verified')
        })
      })
    })
  })

  describe('when it is another user profile', () => {
    let wrapper
    let props
    beforeEach(() => {
      props = {
        ...commonProps,
        match: {
          params: {
            uid: 208
          }
        },
        user: {
          first_name: 'Jane',
          last_name: 'Doe'
        }
      }

      wrapper = shallow(
        <Profile {...props} />
      )
      jest.spyOn(wrapper.instance(), 'notifyFn').mockImplementation(function() {})
    })

    describe('componentWillMount', () => {
      it('calls props.fetchOtherUserProfile', () => {
        expect(props.fetchUserProfile).toHaveBeenCalled()
      })
    })

    describe('filterEntries', () => {
      describe('when uid exists', () => {
        it('returns correct value', () => {
          const item = { id: 1, verified_status: 'declined' }
          wrapper.instance().filterEntries(item)
          expect(item.verified_status).toEqual('not_verified')
        })
      })
    })

    describe('when user profile is not verified', () => {
      it('renders correctly', () => {
        const { store } = config
        const newProps = {
          ...props,
          user: {
            ...props.user,
            biometrics_status: ''
          },
          pending: false
        }
        const tree = renderer.create(
          <Provider store={store}>
            <Router>
              <Profile {...newProps} />
            </Router>
          </Provider>
        ).toJSON()

        expect(tree).toMatchSnapshot()
      })
    })

    describe('when props.userProfileError !== "" ', () => {
      it('renders correctly', () => {
        const { store } = config
        const newProps = {
          ...props,
          user: {
            ...props.user,
            biometrics_status: 'complete'
          },
          pending: false,
          userProfileError: 'profile not found or error'
        }
        const tree = renderer.create(
          <Provider store={store}>
            <Router>
              <Profile {...newProps} />
            </Router>
          </Provider>
        ).toJSON()

        expect(tree).toMatchSnapshot()
      })
    })
  })
})
