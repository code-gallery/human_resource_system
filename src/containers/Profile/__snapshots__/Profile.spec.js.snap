// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Profile when it is another user profile when props.userProfileError !== ""  renders correctly 1`] = `
<div>
  <div
    className="Layout"
  >
    <header
      className="Header"
    >
      <a
        className="Header__appii"
        href="https://appii.io"
        onClick={[Function]}
      >
        <svg
          className=""
          viewBox="0 0 56 56"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M44,156 C28.536027,156 16,143.463973 16,128 C16,112.536027 28.536027,100 44,100 C59.463973,100 72,112.536027 72,128 L72,156 L44,156 Z M44,150.105263 L66.1052632,150.105263 L66.1052632,128 C66.1052632,115.7916 56.2083997,105.894737 44,105.894737 C31.7916003,105.894737 21.8947368,115.7916 21.8947368,128 C21.8947368,140.2084 31.7916003,150.105263 44,150.105263 Z M26.3665673,127.071057 L32.1597485,125.981574 C33.7125958,134.238633 37.761277,138.092554 44.8273599,138.309665 C49.2897633,138.446777 52.4883032,136.287156 54.8218027,132.382701 C55.6587806,130.982254 56.3133349,129.459714 56.7981447,127.935549 C57.0857817,127.031264 57.2482618,126.349163 57.3089183,126.008965 L63.1121343,127.043666 C63.0085691,127.624521 62.7870996,128.554263 62.4155532,129.722346 C61.8017242,131.652128 60.9713937,133.583535 59.8817351,135.406772 C56.5383203,141.001036 51.4867554,144.4118 44.6463244,144.201622 C34.6757726,143.895268 28.4086643,137.929637 26.3665673,127.071057 Z"
            fill="#fff"
            transform="translate(-16 -100)"
          />
        </svg>
      </a>
    </header>
    <div
      className="ContentCard col-xs-12 ContentCard-marginTop"
    >
      <div
        className="ContentCard-wrapper"
      >
        <h3
          className="ContentCard-title"
        >
          Profile not found
        </h3>
        <p>
          Unfortunately, we cannot find the profile you are looking for.
        </p>
      </div>
    </div>
  </div>
  <footer
    className="Footer"
  >
    <div
      className="container"
    >
      <div
        className="row"
      >
        <div
          className="col-xs-12 col-md-4"
        >
          APPII Ltd © 
          2019
        </div>
        <div
          className="col-xs-12 col-md-8"
        >
          <ul
            className="Footer-nav"
          >
            <li>
              <a
                href="/privacy-policy"
                onClick={[Function]}
              >
                Privacy Policy
              </a>
            </li>
            <li>
              <a
                href="/terms-of-use"
                onClick={[Function]}
              >
                Terms & Conditions of Use
              </a>
            </li>
            <li>
              <a
                href="/cookies"
                onClick={[Function]}
              >
                Cookies
              </a>
            </li>
            <li
              className="social-link"
            >
              <a
                href="https://www.linkedin.com/company/10535440/"
              >
                <img
                  alt="linkedin"
                  src="linkedin_logo.svg"
                />
              </a>
            </li>
            <li
              className="social-link"
            >
              <a
                href="https://www.facebook.com/appii.io"
              >
                <img
                  alt="facebook"
                  src="facebook_logo.svg"
                />
              </a>
            </li>
            <li
              className="social-link"
            >
              <a
                href="https://twitter.com/appii_io"
              >
                <img
                  alt="twitter"
                  src="twitter_logo.svg"
                />
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </footer>
</div>
`;

exports[`containers/Profile when it is another user profile when user profile is not verified renders correctly 1`] = `
<div>
  <div
    className="Layout"
  >
    <header
      className="Header"
    >
      <a
        className="Header__appii"
        href="https://appii.io"
        onClick={[Function]}
      >
        <svg
          className=""
          viewBox="0 0 56 56"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M44,156 C28.536027,156 16,143.463973 16,128 C16,112.536027 28.536027,100 44,100 C59.463973,100 72,112.536027 72,128 L72,156 L44,156 Z M44,150.105263 L66.1052632,150.105263 L66.1052632,128 C66.1052632,115.7916 56.2083997,105.894737 44,105.894737 C31.7916003,105.894737 21.8947368,115.7916 21.8947368,128 C21.8947368,140.2084 31.7916003,150.105263 44,150.105263 Z M26.3665673,127.071057 L32.1597485,125.981574 C33.7125958,134.238633 37.761277,138.092554 44.8273599,138.309665 C49.2897633,138.446777 52.4883032,136.287156 54.8218027,132.382701 C55.6587806,130.982254 56.3133349,129.459714 56.7981447,127.935549 C57.0857817,127.031264 57.2482618,126.349163 57.3089183,126.008965 L63.1121343,127.043666 C63.0085691,127.624521 62.7870996,128.554263 62.4155532,129.722346 C61.8017242,131.652128 60.9713937,133.583535 59.8817351,135.406772 C56.5383203,141.001036 51.4867554,144.4118 44.6463244,144.201622 C34.6757726,143.895268 28.4086643,137.929637 26.3665673,127.071057 Z"
            fill="#fff"
            transform="translate(-16 -100)"
          />
        </svg>
      </a>
    </header>
    <div
      className="ContentCard col-xs-12 ContentCard-marginTop"
    >
      <div
        className="ContentCard-wrapper"
      >
        <h3
          className="ContentCard-title"
        >
          Unverified profile
        </h3>
        <p>
          Unfortunately, you cannot view this profile until it has been biometrically identified.
        </p>
      </div>
    </div>
  </div>
  <footer
    className="Footer"
  >
    <div
      className="container"
    >
      <div
        className="row"
      >
        <div
          className="col-xs-12 col-md-4"
        >
          APPII Ltd © 
          2019
        </div>
        <div
          className="col-xs-12 col-md-8"
        >
          <ul
            className="Footer-nav"
          >
            <li>
              <a
                href="/privacy-policy"
                onClick={[Function]}
              >
                Privacy Policy
              </a>
            </li>
            <li>
              <a
                href="/terms-of-use"
                onClick={[Function]}
              >
                Terms & Conditions of Use
              </a>
            </li>
            <li>
              <a
                href="/cookies"
                onClick={[Function]}
              >
                Cookies
              </a>
            </li>
            <li
              className="social-link"
            >
              <a
                href="https://www.linkedin.com/company/10535440/"
              >
                <img
                  alt="linkedin"
                  src="linkedin_logo.svg"
                />
              </a>
            </li>
            <li
              className="social-link"
            >
              <a
                href="https://www.facebook.com/appii.io"
              >
                <img
                  alt="facebook"
                  src="facebook_logo.svg"
                />
              </a>
            </li>
            <li
              className="social-link"
            >
              <a
                href="https://twitter.com/appii_io"
              >
                <img
                  alt="twitter"
                  src="twitter_logo.svg"
                />
              </a>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </footer>
</div>
`;
